os/fs/smartfs/smartfs_utils.c

```diff
@@ -1065,209 +1163,276 @@ int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector,
 		readwrite.buffer = (uint8_t *)fs->fs_rwbuffer;
 		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
 		if (ret < 0) {
-			goto errout;
+			return ret;
 		}
 
 		/* Get the next chained sector */
-
 		chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
 		nextsector = SMARTFS_NEXTSECTOR(chainheader);
 
-		/* Search for an empty entry in this sector */
-
+		/* Search for an available entry in this sector */
 		offset = sizeof(struct smartfs_chain_header_s);
 		entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
-		while (offset + entrysize < readwrite.count) {
-			/* Check if this entry is available */
 
-#ifdef CONFIG_SMARTFS_ALIGNED_ACCESS
-			if ((smartfs_rdle16(&entry->flags) == SMARTFS_ERASEDSTATE_16BIT) || ((smartfs_rdle16(&entry->flags) &
-#else
-			if ((entry->flags == SMARTFS_ERASEDSTATE_16BIT) || ((entry->flags &
-#endif
-					(SMARTFS_DIRENT_EMPTY | SMARTFS_DIRENT_ACTIVE)) == (~SMARTFS_ERASEDSTATE_16BIT & (SMARTFS_DIRENT_EMPTY | SMARTFS_DIRENT_ACTIVE)))) {
-				/* We found an empty entry.  Use it. */
+		while ((offset + entrysize) < readwrite.count) {
+			/* Check if this entry is available */
+			fdbg("\nEntry flags = %x\n", entry->flags);
+			if (IS_EMPTY_ENTRY(entry)) {
+				if (is_empty && entry->flags == SMARTFS_ERASEDSTATE_16BIT) {
+					fdbg("\nEmpty entry is foud\n");
+					*is_empty = TRUE;
+				}
 
-				found = TRUE;
-				break;
+				ret_entry->dsector = psector;
+				ret_entry->doffset = offset;
+				ret = OK;
+				goto errout;
 			}
 
-			/* Not available.  Skip to next entry */
-
+			/* Not an available entry, check the next one */
 			offset += entrysize;
 			entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
 		}
+		ret_entry->dsector = psector;
+		psector = nextsector;
+	}
 
-		/* If we found an entry, stop the search */
-
-		if (found) {
-			break;
-		}
-
-		/* If there are no more sectors, then we need to add one to make
-		 * room for the new entry.
-		 */
-
-		if (nextsector == SMARTFS_ERASEDSTATE_16BIT) {
-			/* Allocate a new sector and chain it to the last one */
-
-			ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
-			if (ret < 0) {
-				goto errout;
-			}
+errout:
+	fdbg("\nret = %d, invalidentry->dsector = %d, invalidentry->doffset = %d\n", ret, ret_entry->dsector, ret_entry->doffset);
+	if (ret != OK) {
+		ret = -ENOENT;
+	}
 
-			nextsector = (uint16_t)ret;
+	if (ret == OK) {
+		RESET_ENTRY_FLAGS(ret_entry);
```

Comment:
hmm can we update all entry flag inside of find or create entry?
I thin both of them should complete operation for new entry....
Please define role of each function first.
1. createentry : Create new entry if....
2. findentry: find...
3. writeentry : write entry data on flash.

So, if you modify some entry data in writeentry, actually it means writeentry does another function's role.

Reply:
Done,
Now writeentry takes care of entry->flags

If the entry is being reused, writeentry will first clear its flags and then set new flags before writing the entry to MTD.